stratum/hal/lib/barefoot/p4runtime_bfrt_translator.h

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef STRATUM_HAL_LIB_BAREFOOT_P4RUNTIME_BFRT_TRANSLATOR_H_
+#define STRATUM_HAL_LIB_BAREFOOT_P4RUNTIME_BFRT_TRANSLATOR_H_
+
+#include <memory>
+#include <string>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/memory/memory.h"
+#include "absl/synchronization/mutex.h"
+#include "p4/v1/p4runtime.pb.h"
+#include "stratum/glue/integral_types.h"
+#include "stratum/glue/status/status.h"
+#include "stratum/glue/status/statusor.h"
+#include "stratum/hal/lib/barefoot/bf_sde_interface.h"
+#include "stratum/hal/lib/common/common.pb.h"
+#include "stratum/hal/lib/common/writer_interface.h"
+#include "stratum/lib/macros.h"
+
+namespace stratum {
+namespace hal {
+namespace barefoot {
+
+class P4RuntimeBfrtTranslator {
+ public:
+  virtual ~P4RuntimeBfrtTranslator() = default;
+  virtual ::util::Status PushChassisConfig(const ChassisConfig& config,
+                                           uint64 node_id)
+      LOCKS_EXCLUDED(lock_);
+  virtual ::util::Status PushForwardingPipelineConfig(
+      const ::p4::config::v1::P4Info& p4info) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::WriteRequest> TranslateWriteRequest(
+      const ::p4::v1::WriteRequest& request) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::ReadRequest> TranslateReadRequest(
+      const ::p4::v1::ReadRequest& request) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::ReadResponse> TranslateReadResponse(
+      const ::p4::v1::ReadResponse& request) LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::StreamMessageRequest>
+  TranslateStreamMessageRequest(const ::p4::v1::StreamMessageRequest& request)
+      LOCKS_EXCLUDED(lock_);
+  virtual ::util::StatusOr<::p4::v1::StreamMessageResponse>
+  TranslateStreamMessageResponse(
+      const ::p4::v1::StreamMessageResponse& response) LOCKS_EXCLUDED(lock_);
+  static std::unique_ptr<P4RuntimeBfrtTranslator> CreateInstance(
+      BfSdeInterface* bf_sde_interface, int device_id,
+      bool translation_enabled) {
+    return absl::WrapUnique(new P4RuntimeBfrtTranslator(
+        bf_sde_interface, device_id, translation_enabled));
+  }
+
+ protected:
+  // Default constructor.
+  P4RuntimeBfrtTranslator()
+      : device_id_(0),
+        bf_sde_interface_(nullptr),
+        translation_enabled_(false),
+        pipeline_require_translation_(false) {}
+
+ private:
+  // Private constructor. Use CreateInstance() to create an instance of this
+  // class.
+  P4RuntimeBfrtTranslator(BfSdeInterface* bf_sde_interface, int device_id,
+                          bool translation_enabled)
+      : device_id_(device_id),
+        bf_sde_interface_(bf_sde_interface),
+        translation_enabled_(translation_enabled),
+        pipeline_require_translation_(false) {}
+  virtual ::util::StatusOr<::p4::v1::Entity> TranslateEntity(
+      const ::p4::v1::Entity& entity, bool to_sdk) SHARED_LOCKS_REQUIRED(lock_);
```

Comment:
Since these functions are private, it's not a huge issue, but https://abseil.io/tips/94 comes to mind.

Reply:
I will add a comment to the parameter since it will be more clear to other developers.

Reply:
comment seems the best, enum would have been overkill probably.